### PR TITLE
[FEATURE] Registry for DataAssistant classes with ability to execute from DataContext by registered name

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* A registry for `DataAssistant` types by "friendly" aliases.
* Implementation of invocation pattern `data_context.assistants.volume.run(batch_request=batch_request)` (where any registered `DataAssistant` name can be used, "volume" serving as name of currently implemented `DataAssistant`).
* Note that both `data_context.assistants.volume.run(batch_request=batch_request)` and its "friendly" alias `data_context.assistants.volume_data_assistant.run(batch_request=batch_request)` are supported.
* Numerous additional unit tests.
* Improvement of existing unit tests.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- GREAT-735/GREAT-736
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/expectations/registry.py [hunk 1]
```diff
@@ -89,10 +90,38 @@ def register_expectation(expectation: Type["Expectation"]) -> None:  # noqa: F82
...
+    logger.debug(
+        f'Registering the declaration of DataAssistant "{data_assistant_type}" took place.'
+    )
+    _registered_data_assistants[data_assistant_type] = data_assistant
+
+
+def get_data_assistant_impl(
+    data_assistant_type: str,
+) -> Type["DataAssistant"]:  # noqa: F821
+    return _registered_data_assistants.get(data_assistant_type)
```
Comment (cdkini) [inline great_expectations/expectations/registry.py]: Is this nullable? If so, we should make the return `Optional`. If not, we should use `[]` to make it very explicit that we're confident we have the value in the dict.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 2]
```diff
@@ -188,6 +208,10 @@ def name(self) -> str:
     def profiler(self) -> BaseRuleBasedProfiler:
         return self._profiler
 
+    @classmethod
+    def is_abstract(cls):
+        return isabstract(cls)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: Please annotate with bool

Comment (cdkini) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: And add a docstring

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 3]
```diff
@@ -24,11 +26,29 @@
...
+
+    def __new__(cls, clsname, bases, attrs):
+        newclass = super().__new__(cls, clsname, bases, attrs)
+        # noinspection PyUnresolvedReferences
+        if not newclass.is_abstract():
+            newclass.data_assistant_type = camel_to_snake(name=clsname)
+            register_data_assistant(data_assistant=newclass)
+
+        return newclass
+
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: Please annotate (as best as you can)

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    DataAssistantDispatcher intercepts requests for "DataAssistant" classes by their registered names and manages their
+    associated "DataAssistantRunner" objects, which process invocations of calls to "DataAssistant" "run()" methods.
+    """
+
+    DATA_ASSISTANT_TYPE_ALIASES: Dict[str, str] = {
+        "volume": "volume_data_assistant",
+    }
+
+    def __init__(self, data_context: "BaseDataContext"):  # noqa: F821
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py]: Please annotate returns for all dunders (even if `None`).

File: tests/test_utils.py [hunk 5]
```diff
@@ -780,26 +769,3 @@ def clean_athena_db(connection_string: str, db_name: str, table_to_keep: str) ->
...
         engine.dispose()
-
-
-# Chetan - 20220426 - This is to be deleted in favor of
-# `set_consistent_seed_within_numeric_metric_range_multi_batch_parameter_builder`
-def set_bootstrap_random_seed_variable(
-    profiler: BaseRuleBasedProfiler,
-    random_seed: int = RANDOM_SEED,
-) -> None:
-    variables_dict: dict
```
Comment (cdkini) [inline tests/test_utils.py:791]: ❤️

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 6]
```diff
@@ -1741,767 +1739,628 @@ def quentin_expected_metrics_by_domain() -> Dict[Domain, Dict[str, Any]]:
...
+    return _profiler_config
+
+
+@pytest.fixture
+def quentin_expected_expectation_suite(
+    quentin_expected_rule_based_profiler_configuration,
+) -> Callable:
+    def _expectation_suite(name: str) -> ExpectationSuite:
+        expected_expect_table_row_count_to_be_between_expectation_configuration: ExpectationConfiguration = ExpectationConfiguration(
             **{
```
Comment (cdkini) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:1882]: Love me some closures ❤️

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class DataAssistantRunner:
+    """
+    DataAssistantRunner processes invocations of calls to "run()" methods of registered "DataAssistant" classes.
+
+    The approach is to instantiate "DataAssistant" class of specified type with "Validator", containing "Batch" objects,
+    specified by "batch_request", loaded into memory.  Then, "DataAssistant.run()" is issued with given directives.
+    """
+
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py:22]: Do we have this in the design or spec anywhere? Do we really need an entirely new class for all this? Would love to get the team's thoughts on ergonomics.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __getattr__(self, name):
+        # Both, full registered "volume_data_assistant" data_assistant_type and alias name are supported for invocation.
+        data_assistant_cls: Type["DataAssistant"] = get_data_assistant_impl(
+            data_assistant_type=name.lower()
+        ) or get_data_assistant_impl(
+            data_assistant_type=DataAssistantDispatcher.DATA_ASSISTANT_TYPE_ALIASES.get(
+                name.lower(), ""
+            )
+        )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py]: Nitpick - Could we split up this logic out of a single statement? There's a lot going on here.

Additionally, I don't think `Type["DataAssistant"]` is accurate if we possibly return `""`

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py (hunk 7 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py:22]: @cdkini This is an internal implementation that makes the required pattern work (per the JIRA item).  This is very akin to `validator.expect_*`.  Thanks.

File: great_expectations/expectations/registry.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/registry.py]: @cdkini Fixed.  Thanks.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini Done -- thanks!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant.py]: @cdkini I added a docstring (committed).  Thanks!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py]: @cdkini Done -- committed -- thanks!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py (hunk 8 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py]: @cdkini Fixed -- thanks!

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 9]
```diff
@@ -2670,7 +2620,26 @@ def test_execution_time_within_proper_bounds(
...
+):
+    context: DataContext = quentin_columnar_table_multi_batch_data_context
+
+    batch_request: dict = {
+        "datasource_name": "taxi_pandas",
+        "data_connector_name": "monthly",
+        "data_asset_name": "my_reports",
+    }
+
+    data_assistant_result: DataAssistantResult = context.assistants.volume.run(
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2634]: :heart:

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 10]
```diff
@@ -2713,7 +2682,39 @@ def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_n
...
+    )
+    assert len(expectation_suite.expectations) == 0
+
+    batch_request: dict = {
+        "datasource_name": "taxi_pandas",
+        "data_connector_name": "monthly",
+        "data_asset_name": "my_reports",
+    }
+
+    data_assistant_result: DataAssistantResult = context.assistants.volume.run(
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2704]: it works!

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 11]
```diff
@@ -2781,7 +2782,69 @@ def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_y
...
+        data_context=context,
+        expectation_suite=None,
+        expectation_suite_name=expectation_suite_name,
+        component_name=None,
+    )
+    assert len(data_assistant_result.expectation_suite.expectations) == 19
+    assert len(expectation_suite.expectations) == 0
+
+    # Using alias name "volume" for invocation.
+    data_assistant_result = context.assistants.volume.run(
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2831]: it's beautiful :)

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 12]
```diff
@@ -2781,7 +2782,69 @@ def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_y
...
+
+    batch_request: dict = {
+        "datasource_name": "taxi_pandas",
+        "data_connector_name": "monthly",
+        "data_asset_name": "my_reports",
+    }
+
+    data_assistant_result: DataAssistantResult
+
+    # Both, full registered "volume_data_assistant" data_assistant_type and alias name are supported for invocation.
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py]: Thank you for this in-line comment. very helpful :)

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 13]
```diff
@@ -2820,34 +2883,91 @@ def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_y
     assert len(expectation_suite.expectations) == 19
 
 
+def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_yes_use_suite_using_implicit_invocation(
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2886]: I love how tight this test this. It's really a testament to how much work is being done under the hood

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 14]
```diff
@@ -2564,6 +2429,20 @@ def run_volume_data_assistant_result_jupyter_notebook_with_new_cell(
...
+    )
+    """
+
+    notebook_code: str
+    if implicit:
+        notebook_code = notebook_code_initialization + implicit_invocation_code
+    else:
+        notebook_code = notebook_code_initialization + explicit_instantiation_code
+
     nb = nbformat.v4.new_notebook()
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2446]: :heart:

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 15]
```diff
@@ -2820,34 +2883,91 @@ def test_volume_data_assistant_add_expectation_configurations_to_suite_inplace_y
...
     bobby_columnar_table_multi_batch_deterministic_data_context,
 ):
     context: DataContext = bobby_columnar_table_multi_batch_deterministic_data_context
+
     new_cell: str = "data_assistant_result.plot()"
+
+    run_volume_data_assistant_result_jupyter_notebook_with_new_cell(
+        context=context,
+        new_cell=new_cell,
+        implicit=False,
```
Comment (Shinnnyshinshin) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2944]: Notes from synchronous review: 
- implicit=False. Will instantiate a Validator, which will instantiate DataAssistant
- implicit=True. Will instatiate DataAssistant from DataContext

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 16]
```diff
@@ -574,6 +574,9 @@ def get_validator_with_expectation_suite(
     Use "expectation_suite" if provided.  If not, then if "expectation_suite_name" is specified, then create
     "ExpectationSuite" from it.  Otherwise, generate temporary "expectation_suite_name" using supplied "component_name".
     """
+    assert expectation_suite is None or isinstance(expectation_suite, ExpectationSuite)
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/helpers/util.py:577]: I like these alot. Thank you @alexsherstinsky

Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/helpers/util.py:577]: I'd love to see more of this in our codebase.

File: great_expectations/data_context/data_context/base_data_context.py [hunk 17]
```diff
@@ -399,6 +402,8 @@ def __init__(
         self._evaluation_parameter_dependencies_compiled = False
         self._evaluation_parameter_dependencies = {}
 
+        self._assistants = DataAssistantDispatcher(data_context=self)
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context/base_data_context.py:405]: :heart:

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(self, data_context: "BaseDataContext") -> None:  # noqa: F821
+        """
+        Args:
+            data_context: BaseDataContext associated with DataAssistantDispatcher
+        """
+        self._data_context = data_context
+
+        self._data_assistant_runner_cache = {}
+
+    def __getattr__(self, name) -> DataAssistantRunner:
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_dispatcher.py:29]: Love it

File: tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py [hunk 19]
```diff
@@ -2608,39 +2488,109 @@ def test_get_metrics_and_expectations(
...
+        "data_asset_name": "my_reports",
+    }
 
+    registered_data_assistant_name: str = "volume_data_assistant"
+
+    expected_expectation_suite: ExpectationSuite = quentin_expected_expectation_suite(
+        name=registered_data_assistant_name
+    )
+
+    data_assistant_result: DataAssistantResult = context.assistants.volume.run(
```
Comment (NathanFarmer) [inline tests/rule_based_profiler/data_assistant/test_volume_data_assistant.py:2558]: 🥇

**Review by NathanFarmer — APPROVED** [review]
LGTM!

File: great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+            data_context=self._data_context,
+            expectation_suite=expectation_suite,
+            expectation_suite_name=expectation_suite_name,
+            component_name=data_assistant_name,
+        )
+        data_assistant: DataAssistant = self._data_assistant_cls(
+            name=data_assistant_name,
+            validator=validator,
+        )
+        data_assistant_result: DataAssistantResult = data_assistant.run(
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant_runner.py:56]: Notes from synchronous review 
- `DataAssistantResult` have `Suite` + Metrics + profiler config + executiontime
- they can also be `.plot()`ed

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 21]
```diff
@@ -188,6 +213,16 @@ def name(self) -> str:
...
 
+    @classmethod
+    def is_abstract(cls) -> bool:
+        """
+        This method inspects the present class and determines whether or not it contains abstract methods.
+
+        Returns:
+            Boolean value (True if all interface methods are implemented; otherwise, False)
+        """
+        return isabstract(cls)
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py:224]: Love this check

**Review by Shinnnyshinshin — APPROVED** [review]
This looks absolutely wonderful Alex. Thank you for this work, and I understand that you will be adding a few docstrings + 1 typofix before merging.

**Pull request merged**